lib/dependencies.js

```diff
@@ -0,0 +1,384 @@
+    let fsTrees = [];
+    for (let fsRoot of this.fsTrees.keys()) {
+      let fsTree = this.fsTrees.get(fsRoot);
+      /** @type {Array<{relativePath: string} & ({type: 'stat', size: number, mtime: number, mode: number} | {type: 'hash', hash: string})>} */
+      let entries = [];
+      for (let entry of fsTree.entries) {
+        if (entry instanceof HashEntry) {
+          entries.push({
+            type: "hash",
+            relativePath: entry.relativePath,
+            hash: entry.hash,
+          });
+        } else {
+          entries.push({
+            type: "stat",
+            relativePath: entry.relativePath,
+            size: entry.size,
+            mtime: +entry.mtime,
+            mode: entry.mode
+          });
+        }
+      }
+      fsTrees.push({
+        fsRoot,
+        entries
+      });
+    }
+    let serialized = {
+      rootDir: this.rootDir,
+      dependencies,
+      fsTrees
+    };
+    return serialized;
+  }
+
+  /**
+   * Deserialize from JSON data returned from the `serialize` method.
+   *
+   * @param dependencyData {ReturnType<Dependencies['serialize']>}
+   * @param [newRootDir] {string | undefined}
+   * @return {Dependencies};
+   */
+  static deserialize(dependencyData, newRootDir) {
+    let oldRootDir = dependencyData.rootDir;
+    newRootDir = newRootDir || oldRootDir;
+    let dependencies = new Dependencies(newRootDir);
+    let files = Object.keys(dependencyData.dependencies);
+    for (let file of files) {
+      let deps = dependencyData.dependencies[file];
+      if (newRootDir) {
+        for (let i = 0; i < deps.length; i++) {
+          let dep = deps[i];
+          if (dep.startsWith(oldRootDir+path.sep)) {
+            deps[i] = dep.replace(oldRootDir, newRootDir);
+          }
+        }
+      }
+      dependencies.setDependencies(file, deps);
+    }
+    /** @type {Map<string, FSTree>} */
+    let fsTrees = new Map();
+    for (let fsTreeData of dependencyData.fsTrees) {
+      /** @type {Array<Entry | HashEntry>} */
+      let entries = [];
+      for (let entry of fsTreeData.entries) {
+        if (entry.type === "stat") {
+          entries.push(new Entry(entry.relativePath, entry.size, entry.mtime, entry.mode));
+        } else {
+          entries.push(new HashEntry(entry.relativePath, entry.hash));
+        }
+      }
+      /** @type {FSTree | FSHashTree}  */
+      let fsTree;
+      /** @type {string} */
+      let treeRoot;
+      if (fsTreeData.fsRoot === oldRootDir) {
+        treeRoot = newRootDir;
+        fsTree = FSHashTree.fromHashEntries(entries, { sortAndExpand: true });
+      } else {
+        treeRoot = fsTreeData.fsRoot;
+        fsTree = FSTree.fromEntries(entries, { sortAndExpand: true });
+      }
+      fsTrees.set(treeRoot, fsTree);
+    }
+    dependencies.seal();
+    dependencies.fsTrees = fsTrees;
+    return dependencies;
+  }
+};
+
+/**
+ * Get an FSTree that uses content hashing information to compare files to
+ * see if they have changed.
```

Comment:
Can you include a brief TL;DR on when in the process we use hashing vs stating. I found myself struggling to understand that while reading the code.